packages/loopring_v3.js/src/request_processors/deposit_processor.ts

```diff
@@ -0,0 +1,71 @@
+import BN from "bn.js";
+import { Bitstream } from "../bitstream";
+import { Constants } from "../constants";
+import { Account, Block, Deposit, ExchangeState } from "../types";
+
+/**
+ * Processes deposit requests.
+ */
+export class DepositProcessor {
+  public static processBlock(state: ExchangeState, block: Block) {
+    const offset = 4 + 32 + 32 + 32 + 32;
+    const data = new Bitstream(block.data);
+    const startIdx = data.extractUint32(offset);
+    const length = data.extractUint32(offset + 4);
+
+    const deposits: Deposit[] = [];
+    for (let i = startIdx; i < startIdx + length; i++) {
+      const deposit = state.deposits[i];
+
+      this.processDeposit(state, deposit);
+
+      deposit.blockIdx = block.blockIdx;
+      deposit.requestIdx = state.processedRequests.length + i - startIdx;
+      deposits.push(deposit);
+    }
+    return deposits;
```

Comment:
I also think this class shall emit events so this class's caller can persist the events and maybe states into a DB o choice. 

In general, these processors CAN maintain in-memory state after loading a state-snapshot from DB upon start/restart, then they should omit new events and new state-values for callers to persist.

Reply:
The database may have two types of tables, one for events:
```
event_type, block_idx, request_idx, ... other data
```
and for each event type a state table 
```
state_type, block_idx ... other accumulative data
```

Event `Deposit`/`Withdrawl`'s corresponding state is `Balance`
Event `Trade`/`Cancel`'s corresponding state is `Order`

---

When we revert block N+1, we delete all events data with block_idx >= N+1 and all states wich block_idx >= N+1,
then we can reload the state with block_idx === N, ....

---

The caller of these processors can keep up to most recent M state records but all event records.

---

The state shall be calculated by one single thread in memory based on the last state and events in the current block. But we probably want to query last state from DB on demand instead of loading all of them in the memory -- it may become too large.

Reply:
I think we can merge the PR, then let me define the DB access API, then you can use the API to load and write data.

Reply:
> I also think this class shall emit events so this class's caller can persist the events and maybe states into a DB o choice.

The code can use an interface + implementation provided by the app for storing/retrieving data. Just an idea instead of having to return the event data all over the place, because the library ideally also has easy access to all state. The implementation can be left to the app.

> In general, these processors CAN maintain in-memory state after loading a state-snapshot from DB upon start/restart, then they should omit new events and new state-values for callers to persist.

Keeping all the accounts in memory + balances + trading history for the current state will take A LOT of memory when there are a lot of accounts. Which could be fine for the server that will use this library for the website. Ideally this lib + database could also be used by normal users so they don't have to depend on a web server to create the Merkle proofs for the withdrawal in withdrawal mode. These users will have a normal PC with a couple of GBs of RAM.

> The database may have two types of tables, one for events:
event_type, block_idx, request_idx, ... other data
and for each event type a state table
state_type, block_idx ... other accumulative data
Event Deposit/Withdrawl's corresponding state is Balance
Event Trade/Cancel's corresponding state is Order

Not really sure what's stored in these tables. Is this comparable to what I do now with the `processedRequests` but specifically for every state value that was updated? The on-chain Deposit/Withdrawal events don't immediately update the state and never have to be reverted, only when they are processed does the state they modified need be reverted, but the event is still valid, it just needs to be processed in a new block. But these events in the table would be 'events' that actually modify the state, not the actual on-chain events?

> When we revert block N+1, we delete all events data with block_idx >= N+1 and all states wich block_idx >= N+1,
then we can reload the state with block_idx === N, ....

How can we reload the state with block_idx? Do we revert the state updates stored in the state events on the current block until we get to the block we reverted to?

> The state shall be calculated by one single thread in memory based on the last state and events in the current block. But we probably want to query last state from DB on demand instead of loading all of them in the memory -- it may become too large.

Yes, that's how I would do it as well. Just load the account from DB when needed.

> I think we can merge the PR, then let me define the DB access API, then you can use the API to load and write data.

I think that's a good idea. It should be straightforward for me to update the code where necessary.